### PR TITLE
Feature/cbs pfpksk serde

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves:

closes https://github.com/zama-ai/concrete-core-internal/issues/396

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [x] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-core-ffi/tests/test_lwe_bit_extraction.c [hunk 1]
```diff
@@ -602,7 +602,7 @@ void extract_bits_unchecked_raw_ptr_buffers_test(void) {
     uint64_t expected = (val >> (number_of_bits_of_message - 1 - idx)) & 1;
     uint64_t rounded = closest_representable(output_plaintext_buffer[idx], 1, 1);
     uint64_t decrypted = rounded >> 63;
-    printf("decrypted %ld, expected %ld\n", decrypted, expected);
+    printf("decrypted %" PRIu64 ", expected %" PRIu64 "\n", decrypted, expected);
```
Comment (agnesLeroy) [inline concrete-core-ffi/tests/test_lwe_bit_extraction.c:605]: Why these changes?

File: concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_deserialization.rs [hunk 2]
```diff
@@ -1683,6 +1688,194 @@ impl EntityDeserializationEngine<&[u8], LweCiphertextVector64> for DefaultSerial
...
+                .map_err(DefaultSerializationError::Deserialization)
+                .map_err(EntityDeserializationError::Engine)?;
+        match deserialized {
+            DeserializableLweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys64 {
+                version:
+                    LweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys64Version::Unsupported,
+                ..
+            } => Err(EntityDeserializationError::Engine(
+                DefaultSerializationError::UnsupportedVersion,
+            )),
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_deserialization.rs:1861]: Just reading this code I'm thinking that as soon as we have more serialization versions this will be a pain to maintain :sob:

File: concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs [hunk 3]
```diff
@@ -1878,6 +1883,164 @@ impl<'b> EntitySerializationEngine<LweCiphertextMutView64<'b>, Vec<u8>>
...
+        bincode::serialize(&serializable)
+            .map_err(DefaultSerializationError::Serialization)
+            .map_err(EntitySerializationError::Engine)
+    }
+
+    unsafe fn serialize_unchecked(
+        &mut self,
+        entity: &LweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys32,
+    ) -> Vec<u8> {
+        self.serialize(entity).unwrap()
```
Comment (agnesLeroy) [inline concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs:1961]: Here basically the unchecked version is useless? Why provide a serialization function that does not serialize the version?

**Review by agnesLeroy — APPROVED** [review]
Hey again! This one looks good to me, you'll see I have a few questions but not blocking  :slightly_smiling_face:

File: concrete-core-ffi/tests/test_lwe_bit_extraction.c (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core-ffi/tests/test_lwe_bit_extraction.c:605]: for M1 mac, basically compilers don't agree on what a uint64_t type is for printing so they came up with a macro that does just that https://stackoverflow.com/questions/16859500/what-is-priu64-in-c

File: concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_deserialization.rs (hunk 2 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_deserialization.rs:1861]: yep, we'll take the time for engineering this, no worries

File: concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs:1961]: hmm what do you mean ? It goes through the serialize function just above ?

Comment (agnesLeroy) [reply concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs:1961]: Oh yes :sweat_smile:  so really this is useless?

Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs:1961]: well kind of 😅 it was a discussion with Alex at the time

Comment (agnesLeroy) [reply concrete-core/src/backends/default/implementation/engines/default_serialization_engine/entity_serialization.rs:1961]: well if we drop support for the unchecked functions this question will disappear :sweat_smile:
